index.test-d.ts

```diff
@@ -0,0 +1,65 @@
+import {expectType, expectError, expectAssignable} from 'tsd';
+import pify = require('.');
+
+const fs = {
+    readFile(path: string, callback: (error?: Error, data?: Buffer) => void): void {
+        callback(undefined, new Buffer('abc'))
+    },
+    exists(path: string, callback: (result: boolean) => void): void {
+        callback(true)
+    }
+}
+
+// other types
+expectError(pify());
+expectError(pify(42));
+expectError(pify('abc'));
+expectType<number[]>(pify([1,2]));
+
+// functions
+expectAssignable<() => Promise<never>>(pify(() => {}))
+expectType<Promise<never>>(pify(() => {})())
+
+expectAssignable<(path: string) => Promise<Buffer | undefined>>(pify(fs.readFile))
+expectType<Promise<Buffer | undefined>>(pify(fs.readFile)('a.txt'))
+
+declare function request(
+    url: string,
+    callback: (error?: Error, response?: { statusCode: number }, body?: string) => void
+): void;
+expectAssignable<(url: string) => Promise<[{ statusCode: number }?, string?]>>(pify(request, { multiArgs: true }))
+expectType<Promise<[{ statusCode: number }?, string?]>>(pify(request, { multiArgs: true })('https://www.google.com'))
+
+declare function request2(
+    url: string,
+    callback: (response?: { statusCode: number }, body?: string) => void
+): void;
+expectAssignable<(url: string) => Promise<[{ statusCode: number }?, string?]>>(pify(request2, { multiArgs: true, errorFirst: false }))
+expectType<Promise<[{ statusCode: number }?, string?]>>(pify(request2, { multiArgs: true, errorFirst: false })('https://www.google.com'))
+
+expectAssignable<(path: string) => Promise<boolean>>(pify(fs.exists, { errorFirst: false }))
+expectType<Promise<boolean>>(pify(fs.exists, { errorFirst: false })('a.txt'))
+
+expectAssignable<(path: string) => Promise<Buffer | undefined>>(pify(fs.readFile, { promiseModule: Promise }))
+expectType<Promise<Buffer | undefined>>(pify(fs.readFile, { promiseModule: Promise })('a.txt'))
+
+declare function generic<A, B>(arg1: A, arg2: B, callback: (error?: Error, result?: A[]) => void): void
+expectAssignable<(arg1: number, arg2: string) => Promise<number[] | undefined>>(pify<(arg1: number, arg2: string, callback: (error?: Error, result?: number[]) => void) => void>(generic))
+expectType<Promise<number[] | undefined>>(pify<(arg1: number, arg2: string, callback: (error?: Error, result?: number[]) => void) => void>(generic)(123, 'abc'))
```

Comment:
I didn't find a way to make pify returns a generic function again. Users have to specify types of `PifyInput` explicitly. 😢

Reply:
I did warn people in the original issue:

> **Note that this is not an easy task and requires you to have advanced TypeScript experience.**

😛